### PR TITLE
HPCC-30900 Add ListWU to ESP WSSasha service

### PR DESCRIPTION
**Pull request by wangkx**
The ListWU has the similar functions to Sasha CLI's List action and ListDT action. It can be used to retrieve the information about archived WUs with filters, such as cluster, state, etc.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dali/sasha/saarch.cpp [hunk 1]
```diff
@@ -540,6 +542,11 @@ void WUiterate(ISashaCommand *cmd, const char *mask)
                         }
                     }
 
+                    if (wuCount < wuStart)
```
Comment (jakesmith) [inline dali/sasha/saarch.cpp:545]: I'm not convinced the wuStart logic here is right, or necessary.
Paging was achecived before by passing in the last workunit as a before or after in the request.

The count should always be the number you want to return (not start+count).

File: dali/sasha/saruncmd.cpp [hunk 2]
```diff
@@ -46,6 +47,61 @@ class CSashaCmdExecutor : public CInterfaceOf<ISashaCmdExecutor>
...
+        if (!req->state.isEmpty())
+            cmd->setState(req->state.get());
+
+        if (!req->fromDate.isEmpty())
+            cmd->setAfter(req->fromDate.str()); //In CArchivedWUReader, the After is used as fromDate.
+        if (!req->toDate.isEmpty())
+            cmd->setBefore(req->toDate.str()); //In CArchivedWUReader, the Before is used as toDate.
+
+        cmd->setStart(req->pageStartFrom); //In CArchivedWUReader, the Start is the index of the first item to be returned.
+        cmd->setLimit(req->pageStartFrom + req->pageSize); //In CArchivedWUReader, the Limit is the index after the last item to be returned.
```
Comment (jakesmith) [inline dali/sasha/saruncmd.cpp:78]: see comment in saarch.cpp.
At the moment, I don't think a pageStartFrom should be part of the request, but rather paging should use the previous approach of specifying the beforeWu/afterWu.
And the count/limit should be an absolute number, i.e. not offset+count.

**Review by jakesmith — COMMENTED** [review]
@wangkx - looks good in general, but some questions re. introduction of a start offset.

**Comment by wangkx** [comment]
@jakesmith I may use the beforeWu/afterWu for paging. But, I am not sure that ECLWatch UI code would like that because many other functions (non-archived WUs, logical files, etc) are using PageStartFrom and PageSize for paging.

**Comment by jakesmith** [comment]
>I may use the beforeWu/afterWu for paging. But, I am not sure that ECLWatch UI code would like that because many other functions (non-archived WUs, logical files, etc) are using PageStartFrom and PageSize for paging.

@wangkx - other solutions use a cache to get all results, then allow paging through it with an offset (page-start). The cache minimizes the cost of a subsequent requests from the same client paging through results (since it has the next results cached).

We have to be very careful not to add functionality that is prohibitively expensive each time it is used.
If Each time it had to get results from Sasha, it had to walk through all directories reading files to get to the offset that would be bad (a 'feature' we would not want to introduce).
NB: there are comments elsewhere re. this issue:
```
void doWUQueryFromArchive(IEspContext &context, const char* sashaServerIP, unsigned sashaServerPort,
       ArchivedWuCache &archivedWuCache, unsigned cacheMinutes, IEspWUQueryRequest & req, IEspWUQueryResponse & resp)
{
    //Sasha server does noy support the PageStartFrom due to inefficient access to archived workunits for pages>1.
```

If the Eclwatch UI code can't use the last item in the previous list to request the next page. It may be best not to support paging of archived workunits at all.

**Comment by wangkx** [comment]
@jakesmith the PageStartFrom/PageSize are replaced by the beforeWU/afterWU. Please review.

**Review by jakesmith — COMMENTED** [review]
@wangkx - looks good. Please squash

**Comment by wangkx** [comment]
@jakesmith I squashed.

**Review by jakesmith — APPROVED** [review]
@wangkx - looks good.

**Comment by wangkx** [comment]
@ghalliday please merge

**Pull request merged**